src/vm/methodtable.cpp

```diff
@@ -7171,43 +7174,62 @@ BOOL MethodTable::FindDefaultInterfaceImplementation(
-                                        continue;
+                                // Is this the right slot?
+                                if (pDeclMD->GetSlot() != targetSlot)
+                                    continue;
 
-                                    MethodTable *pDeclMT = pDeclMD->GetMethodTable();
-                                    if (pDeclMT->ContainsGenericVariables())
-                                    {
-                                        TypeHandle thInstDeclMT = ClassLoader::LoadGenericInstantiationThrowing(
-                                            pDeclMT->GetModule(),
-                                            pDeclMT->GetCl(),
-                                            pCurMT->GetInstantiation());
-                                        MethodTable *pInstDeclMT = thInstDeclMT.GetMethodTable();
-                                        if (pInstDeclMT == pInterfaceMT)
-                                        {
-                                            // This is a matching override. We'll instantiate pCurMD later
-                                            pCurMD = pMD;
-                                            break;
-                                        }
-                                    }
-                                    else if (pDeclMD == pInterfaceMD)
+                                // Is this the right interface?
+                                if (!pDeclMD->HasSameMethodDefAs(pInterfaceMD))
+                                    continue;
+
+                                if (pInterfaceMD->HasClassInstantiation())
+                                {
+                                    // pInterfaceMD will be in the canonical form, so we need to check the specific
+                                    // instantiation against pInterfaceMT.
+                                    //
+                                    // The parent of pDeclMD is unreliable for this purpose because it may or
+                                    // may not be canonicalized. Let's go from the metadata.
+
+                                    SigTypeContext typeContext = SigTypeContext(pCurMT);
+
+                                    mdTypeRef tkParent;
+                                    IfFailThrow(pMD->GetModule()->GetMDImport()->GetParentToken(it.GetToken(), &tkParent));
+
+                                    MethodTable* pDeclMT = ClassLoader::LoadTypeDefOrRefOrSpecThrowing(
+                                        pMD->GetModule(),
+                                        tkParent,
+                                        &typeContext).AsMethodTable();
+
+                                    // We do CanCastToInterface to also cover variance.
+                                    // We already know this is a method on the same type definition as the (generic)
+                                    // interface but we need to make sure the instantiations match.
```

Comment:
Variant dispatch is typically done via a two pass technique. Wherein we first iterate to see if there is an exact match, and then a second iteration through the interfaces (of a particular type) to attempt to do a variant dispatch. Looking at the logic in this function, it appears that the variant matching rules here may not be followed, and the variance test cases added do not appear to cover the exact/inexact matching issue.

Reply:
The issue of it ignoring variant interface dispatch rules and not doing two passes is orthogonal to the thing I'm fixing here.

I would prefer to track that in a separate issue/pull request for accounting purposes.

Reply:
#20452